### DISCUSSION TITLE
Rework logs

### DISCUSSION THREAD

**Discussion by yohannlog** [coollabsio/coolify#2618] [Improvement Requests]
**1. Logs color**

By default, 'docker logs' command don't support color by default.
I know that docker compose offers a 'tty' option to have the colors support , but this requires changing all docker-compose for supporting it.

**2. Filter by dates**

By default, time is added directly in the logs. 'docker logs'
I want to change the time for a more 'readable' way (Like 'Jun 20  14:10:10' with an hover to add more informations like timezone, etc...) and be able to sort the logs by date and time (by clicking it for example). The problem is that this requires me to process each line to manage the apparent times, etc... This risks impacting performance on large quantities of logs, especially with filters and line limitations.

(For logs and dates): My idea would have been to have an integrated service dedicated to logs. If we don't want to add a new service for that, for dates we can use '--since' parameter (with bash) for the date filter, however, this does not remove the 'workload' of adding a click event on the time or limitations about the command. For colors, we have to add parameters to docker-compose and see what it does.

**3. Tabs for multiple container in one menu**

When you jave, for example, a supabase service, you have a huge quantity of log 'card':
![image](https://github.com/coollabsio/coolify/assets/8901733/d65b20fc-de1c-4d17-91e2-03099d76c563)

My idea is to create tabs for each of them.

**4. Filters and download button**

I've added a 'download logs' button but I'm wondering how to manage it when certain filters are activated and the number of lines. Should I download all the logs by default or should I take all the filters by default?

**5. Raising line limits**

The idea is to set a default limit and add a "show more" button on top if you want to see older logs. After that, we could imagine two buttons at the bottom right of the \<pre> to go to the very top of the logs or to the very bottom. However, I don't know if 'docker logs' can handle something this complex. (That's why we've come up with the idea of an integrated service.)

**6. Parsing the json logs format and proposing a way to 'unfold' it**

What's your opinion ?

**7. What about the hours already integrated by the services themselves?**

If you have any opinions or suggestions, I'd love to hear them. If you have any additions you'd like to make, please feel free to discuss them here.

Thank you !

**Comment by peaklabs-dev** [comment]
@yohannlog 
1. Just leave the log color in this case. As tty adds a security risk because containers do not run rootless, so privilege escalation could be possible -> I would not suggest enabling tty.
2. Personally, I would not add such an option, especially if it reduces performance significantly (as performance is more important than dates on the logs). But this could be an optional option that can be turned on or off, so for small logs I could choose to turn it on and for large logs I could choose to turn it off. Also, we can choose how to display the date: y:m:d -> or d:m:y...
3. Yes tabs would be great
4. Add a checkbox Download filtered logs and if download is pressed the full log will be downloaded and if the checkbox is checked only the filtered log will be downloaded.
5. Show more button sounds great.
6. sounds good
7. I don't think hours that are already logged by the service can be removed they will always be there I believe (and it is fine, only with performance overhead these could be removed which is unnecessary)
8. I think for proper times and dates in the log there should be a server wide timezone setting that just sets the servers timezone via dropdown.

**Reply by yohannlog** [reply]
I've had time to think about this PR. I'll also take into account the opinions. I'll rewrite the PR later today to announce all the big and small changes. We're going to start with a global overhaul of the logs.

**Comment by yohannlog** [comment]
[PR #2598  UPDATED]

**Comment by LEstradioto** [comment]
Nice! The #2651 would benefit from this
I also appreciate your calls.

One thing that could really help with debugging community issues (which are growing rapidly 🚀) is the ability to easily download main logs from unmanaged Docker containers (like Coolify, servers, etc.). This would be very useful for support.

I haven't tested the Drain Logs feature, so I'm not sure if it already captures these unmanaged logs.

**Comment by peaklabs-dev** [comment]
Some more changes to improve logs:

- Add API support
- Send logs to this service
- Save logs in files ( {container_id}/{dd_mm_yy.csv} ) with rotation and file size limits
- Save logs in database
- Send logs to custom service ?
- Add a page dedicated to logs with history (New button to left menu)
- Add "Show more" buttons on large logs
- Supports ANSI color format
- Add tabs when multiple servers in one page
- Add a "Download button"
- Add a checkbox to download using max lines
- Add type="number" to max lines input
- Add search filter
- Add option to download logs on only filtered logs
- Add styles to logs ?
- Add a wrap lines checkbox so that log lines are automatically wrapped (to viewport size).
- Save wrap lines option
- Add style for + and - button (For wrapping lines)
- Make time clickable to filter logs
- Review the code in GetLogs.php to simplify it (duplicate code, etc.).
- Manage json logs format
- Add colors to logs (Red = Error, Yellow = Warning ...):
![image](https://github.com/user-attachments/assets/5cb3669a-0ec6-4be9-9b9f-2e33ab7d7b45)

**Comment by FTCHD** [comment]
+1 here for API support for logs, either streamed or plain GETs like Portainer

**Comment by renequez** [comment]
I believe maybe instead of focusing on reworking logs, just give the Drain Logs feature more love:
1. Better docs, especially with examples for fluentbit
2. Integrate more third party providers like Cloud Grafana, GCP Logging, Datadog etc.
3. One click integration with a Coolify service like Loki, or at least a guide in the docs

Number 3 would be absolutely amazing!
I think this would be the better approach because observability features like logging should maybe not be in the scope of Coolify, but instead Coolify *seamlessly* integrate with specialised platforms.

**Comment by FTCHD** [comment]
API support is a must IMO, any serious solution has it, even the "bad" ones.

at this point I'd say just give us *something*
